fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/Operator.java

```diff
@@ -53,6 +53,15 @@ public abstract class Operator {
     // or self reference of groups
     protected long salt = 0;
 
+    // Like LogicalJoinOperator#transformMask, add a mask to avoid one operator's dead-loop in one transform rule.
+    // eg: MV's UNION-ALL RULE:
+    //                 UNION                         UNION
+    //               /        \                    /       \
+    //  OP -->   EXTRA-OP    MV-SCAN  -->     UNION    MV-SCAN     ---> ....
+    //                                       /      \
+    //                                  EXTRA-OP    MV-SCAN
+    protected int opRuleMask = 0;
```

Comment:
is it necessary to consider it in hashCode and equals?

Reply:
No. It's just a flag to mark the Operator used some rules and so can shortcut, and should not affect `Equality`.

Reply:
This happens when eager union all is set on. I have added a comment below:
```
        // 2. filter out predicates whose column refs are in query's output columns because we can always compensate
                    // by output column using Union-All.
                    // eg:
                    //    query: select dt, k2 from tblA where date_trunc('day', dt) >= '2023-11-01'
                    //      mv : select dt, k2 from tblA where dt = '2023-11-01'
                    // NOTE: This is not supported by default because it can generate some bad cases (not the best plan):
                    // eg:
                    //    query: select dt, k2 from tblA where a < 1
                    //    mv   : select dt, k2 from tblA where a > 2
                    // this rule still can rewrite query by mv like this:
                    //    select * from (select dt, k2 from tblA) t where t.a < 1
                    // rewrite to:
                    //    select * from (select * from mv union all select dt, k2 from tblA where a <= 2 or a is null)
                    //          t where t.a < 1
                    // If you want to use this feature, you can use it by the setting:
                    //      set query_debug_options = "{'enableMVEagerUnionAllRewrite':true}";
                    if (queryOutputColumnRefs != null && queryOutputColumnRefs.contains(col)) {
                        return true;
                    }
```